Add docs about translation handling (#6405)
woocommerce/woocommerce-blocks: docs/translations/translation-basics.md

> # Translation basics

Reviewer: I'm not sure if it's fully worth it or not, but maybe also talking about how translation is shaped, it comes in 3 forms:
- A `.pot` file that contains all the original translation and paths to it, this file is unique to the plugin, for example: `woo-gutenberg-products-block.pot`
- A `.po` and `.mo` which pairs of strings and translation, this is unique to the plugin and language and handles only PHP strings, e.g. `woo-gutenberg-products-block-nl_NL.po`.
- JSON files, which are unique to the plugin langauge, and specific file, those have the [JED](https://messageformat.github.io/Jed/) format and are named as `{plugin_slug}-{language_code}-{md5(filepath)}.json`, for example `woo-gutenberg-products-block-nl_NL-df69ae21391b08cd8133e643dacbc72d.json` is the JSON translation file of `build/checkout-frontend.js` file in Dutch.

Reviewer 2: - We should also explain the loading order of the translation file.
- For JS files need [registering script translation](https://github.com/woocommerce/woocommerce-blocks/blob/5b980254dfbbd4f11f03c8a8b0830c9f738e03ca/src/Assets/Api.php#L161) which is worth noting here IMO.

Author: Thanks for pointing out these cases, @senadir and @dinhtungdu. I added the new doc `docs/translations/translation-loading.md` where I've added these sections.

## Localization functions

> Since [WordPress 2.1 "Ella"](https://wordpress.org/support/wordpress-version/version-2-1/), WordPress offers i18n for PHP files, and since [WordPress 5.0 "Bebo"](https://wordpress.org/support/wordpress-version/version-5-0/), it also offers i18n for JS files. Handling translations is pretty straight forward. Both PHP and JS handle translations similar. WordPress offers the functions:

Reviewer 3: I know i18n is a well-known abbreviation, but maybe we could write `internationalization (i18n)` just in case someone is unfamiliar with the abbreviation?

Author: Great suggestion. I've changed that part accordingly.

-   `__()` → Available in PHP & JS/TS.
-   `_e()` → Available in PHP only.
-   `_ex()` → Available in PHP only.
-   `_n()` → Available in PHP & JS/TS.
-   `_x()` → Available in PHP & JS/TS.
-   `_nx()` → Available in PHP & JS/TS.
-   `esc_html__()` → Available in PHP only.
-   `esc_html_e()` → Available in PHP only.
-   `esc_html_x()` → Available in PHP only.
-   `esc_attr__()` → Available in PHP only.
-   `esc_attr_e()` → Available in PHP only.

> -   `esc_attr_x()` → Available in PHP only.

Reviewer 3: What do you think about linking to these functions on the WordPress code reference site?

Author: In a previous version, I actually linked these functions to their corresponding reference. I am not sure why I removed these links, but having them is handy. I just added them again.

## GlotPress

> All translations are handeled using [GlotPress](https://wordpress.org/plugins/glotpress/). As the WooCommerce Blocks plugin is hosted on https://wordpress.org/, all plugin-related translations can be found and managed on https://translate.wordpress.org/projects/wp-plugins/woo-gutenberg-products-block/.

Reviewer 3: ```suggestion
All translations are handled using [GlotPress](https://wordpress.org/plugins/glotpress/). As the WooCommerce Blocks plugin is hosted on https://wordpress.org/, all plugin-related translations can be found and managed on https://translate.wordpress.org/projects/wp-plugins/woo-gutenberg-products-block/.
```

## Text domain

> Prior to [WordPress 4.6 “Pepper Adams”](https://wordpress.org/support/wordpress-version/version-4-6/), a text domain had to be defined to make the strings translatable. While it’s no longer required having a text domain, it does not harm if it’s still included. If the text domain is available, it has to match the slug of the plugin and is defined in the header of the main plugin file `woocommerce-gutenberg-products-block.php`:

Reviewer 3: ```suggestion
Prior to [WordPress 4.6 “Pepper Adams”](https://wordpress.org/support/wordpress-version/version-4-6/), a text domain had to be defined to make the strings translatable. While it’s no longer a requirement to have a text domain, it does no harm to still include it. If the text domain is available, it has to match the slug of the plugin and is defined in the header of the main plugin file `woocommerce-gutenberg-products-block.php`:
```
